### ISSUE TITLE
Unable to build documentation from "master" branch

### ISSUE THREAD

**Issue by SchweinDeBurg** [fish-shell/fish-shell#5808]
```
$ uname -a
Linux hatrack.spbtv.com 4.16.7-100.fc26.x86_64 #1 SMP Wed May 2 21:45:56 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
$ echo $TERM
xterm-256color
```
I've just tried to build fish from the "master" branch and got the following error:
```
[ 93%] Building HTML documentation with Sphinx

Encoding error:
'ascii' codec can't encode character u'\u2192' in position 31: ordinal not in range(128)
The full traceback has been saved in /home/elijah/.local/tmp/sphinx-err-tUOL2v.log, if you want to report the issue to the developers.
make[2]: *** [CMakeFiles/sphinx-docs.dir/build.make:59: CMakeFiles/sphinx-docs] Error 1
make[2]: Leaving directory '/home/elijah/Laptop/home/rpmbuild/BUILD/fish/3.14.1-x86_64-redhat-linux-gnu'
make[1]: *** [CMakeFiles/Makefile2:401: CMakeFiles/sphinx-docs.dir/all] Error 2
make[1]: Leaving directory '/home/elijah/Laptop/home/rpmbuild/BUILD/fish/3.14.1-x86_64-redhat-linux-gnu'
make: *** [Makefile:130: all] Error 2
make: Leaving directory '/home/elijah/Laptop/home/rpmbuild/BUILD/fish/3.14.1-x86_64-redhat-linux-gnu'
```
Below is the /home/elijah/.local/tmp/sphinx-err-tUOL2v.log:
```
# Sphinx version: 1.5.6
# Python version: 2.7.14 (CPython)
# Docutils version: 0.13.1 release
# Jinja2 version: 2.9.6
# Last messages:
#   writing output... [ 90%] cmds/umask
#   writing output... [ 91%] cmds/vared
#   writing output... [ 92%] cmds/wait
#   writing output... [ 93%] cmds/while
#   writing output... [ 94%] commands
#   writing output... [ 95%] design
#   writing output... [ 96%] faq
#   writing output... [ 97%] index
#   writing output... [ 98%] license
#   writing output... [100%] tutorial
# Loaded extensions:
#   alabaster (0.7.9) from /usr/lib/python2.7/site-packages/alabaster/__init__.pyc
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/cmdline.py", line 296, in main
    app.build(opts.force_all, filenames)
  File "/usr/lib/python2.7/site-packages/sphinx/application.py", line 333, in build
    self.builder.build_update()
  File "/usr/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 251, in build_update
    'out of date' % len(to_build))
  File "/usr/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 322, in build
    self.write(docnames, list(updated_docnames), method)
  File "/usr/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 360, in write
    self._write_serial(sorted(docnames), warnings)
  File "/usr/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 368, in _write_serial
    self.write_doc(docname, doctree)
  File "/usr/lib/python2.7/site-packages/sphinx/builders/html.py", line 452, in write_doc
    self.docwriter.write(doctree, destination)
  File "/usr/lib/python2.7/site-packages/docutils/writers/__init__.py", line 80, in write
    self.translate()
  File "/usr/lib/python2.7/site-packages/sphinx/writers/html.py", line 48, in translate
    self.document.walkabout(visitor)
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 174, in walkabout
    if child.walkabout(visitor):
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 166, in walkabout
    visitor.dispatch_visit(self)
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 1882, in dispatch_visit
    return method(node)
  File "/usr/lib/python2.7/site-packages/sphinx/writers/html.py", line 789, in visit_literal_block
    HTMLTranslator.visit_literal_block(self, node)
  File "/usr/lib/python2.7/site-packages/sphinx/writers/html.py", line 392, in visit_literal_block
    **highlight_args)
  File "/usr/lib/python2.7/site-packages/sphinx/highlighting.py", line 138, in highlight_block
    hlsource = highlight(source, lexer, formatter)
  File "/usr/lib/python2.7/site-packages/pygments/__init__.py", line 85, in highlight
    return format(lex(code, lexer), formatter, outfile)
  File "/usr/lib/python2.7/site-packages/pygments/__init__.py", line 64, in format
    formatter.format(tokens, realoutfile)
  File "/usr/lib/python2.7/site-packages/pygments/formatter.py", line 95, in format
    return self.format_unencoded(tokensource, outfile)
  File "/usr/lib/python2.7/site-packages/pygments/formatters/html.py", line 850, in format_unencoded
    for t, piece in source:
  File "/usr/lib/python2.7/site-packages/pygments/formatters/html.py", line 690, in _wrap_div
    for tup in inner:
  File "/usr/lib/python2.7/site-packages/pygments/formatters/html.py", line 708, in _wrap_pre
    for tup in inner:
  File "/usr/lib/python2.7/site-packages/pygments/formatters/html.py", line 727, in _format_lines
    for ttype, value in tokensource:
  File "/usr/lib/python2.7/site-packages/pygments/lexer.py", line 191, in streamer
    for _, t, v in self.get_tokens_unprocessed(text):
  File "<string>", line 129, in get_tokens_unprocessed
  File "<string>", line 94, in tokenize_fish_command
  File "/usr/lib64/python2.7/subprocess.py", line 479, in communicate
    return self._communicate(input)
  File "/usr/lib64/python2.7/subprocess.py", line 1102, in _communicate
    stdout, stderr = self._communicate_with_poll(input)
  File "/usr/lib64/python2.7/subprocess.py", line 1166, in _communicate_with_poll
    input_offset += os.write(fd, chunk)
UnicodeEncodeError: 'ascii' codec can't encode character u'\u2192' in position 31: ordinal not in range(128)
```

**Comment by zanchey** [comment]
Yes, I get this on Debian using Sphinx (sphinx-build) 1.4.9.

**Comment by faho** [comment]
Those are `→` literals, which we use for arrow keys.

Is a python3 build of sphinx available for debian? Or would you have to install it via `pip`?

**Comment by Darkshadow2** [comment]
Yeah I got this on macOS as well, using the system Python 2.7. I had to install Python 3 to get it to work.

**Comment by mqudsi** [comment]
Can we detect the python2 sphinx and error out instead of shelling out to it (in CMake)?

It's a dev requirement (not runtime) so there's no reason to force ourselves to use shims or other workarounds for python2's broken unicode support (wasn't proper Unicode support a primary driver behind the migration to python3 in the first place?)

**Comment by SchweinDeBurg** [comment]
Package 'python3-sphinx' is available for Fedora. F26 has version 1.5.6, F27 - 1.6.6, F28 - 1.7.2, F29 - 1.7.6 and Rawhide (future F30) - 2.0.0 beta2. So it will be great to use python3 version when building fish.

**Comment by SchweinDeBurg** [comment]
After the following local commit
```
commit 6e7963c4c22007081da4848cda0752741067a835
Refs: [master], 3.0.2-889-g6e7963c4
Author:     Elijah Zarezky <elijah@zarezky.spb.ru>
AuthorDate: Fri Apr 12 06:53:05 2019 +0300
Commit:     Elijah Zarezky <elijah@zarezky.spb.ru>
CommitDate: Fri Apr 12 06:53:05 2019 +0300

    docs: Use Python3 version of sphinx-build
---
 cmake/Docs.cmake | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/cmake/Docs.cmake b/cmake/Docs.cmake
index 9e2758b1..ea1d9c83 100644
--- a/cmake/Docs.cmake
+++ b/cmake/Docs.cmake
@@ -1,4 +1,4 @@
-FIND_PROGRAM(SPHINX_EXECUTABLE NAMES sphinx-build
+FIND_PROGRAM(SPHINX_EXECUTABLE NAMES sphinx-build-3
     HINTS
     $ENV{SPHINX_DIR}
     PATH_SUFFIXES bin
@@ -43,7 +43,7 @@ ELSE(SPHINX_EXECUTABLE)
 ENDIF(SPHINX_EXECUTABLE)

 IF(BUILD_DOCS AND NOT SPHINX_EXECUTABLE)
-    MESSAGE(FATAL_ERROR "build documentation selected, but sphinx-build could not be found")
+    MESSAGE(FATAL_ERROR "build documentation selected, but sphinx-build-3 could not be found")
 ENDIF()

 IF(IS_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}/user_doc/html
```
I've just successfully build fish from its "master" branch.

**Comment by faho** [comment]
So, as it turns out, this is an issue only inside code blocks, which might mean it's in our custom lexer.

But since we only use it twice and it's not even really necessary, let's just remove those uses.

**Comment by faho** [comment]
And, for the record, demanding a python3-using-sphinx would be annoying - my system has _just_ `sphinx-build`, while Fedora apparently has _just_ `sphinx-build-3` (whereas mine has `sphinx-build2` for the python2 version).

**Comment by SchweinDeBurg** [comment]
My system has both: sphinx-build (Python2) and sphinx-build-3 (Python3), they are from different packages, installed into the different places.

**Comment by faho** [comment]
>My system has both: sphinx-build (Python2) and sphinx-build-3 (Python3)

Yes. When I said "my system has just sphinx-build", what I meant was that that was the sole name for the python3 version. The names aren't standardized, so we can't really use them.

`sphinx-build` could be either, and the name for the one with 3 could end in `-3` or just `3`, or it could be called "banana" for all I know. There is no good way to figure this out via the name alone, which is why it's best we work around the issues with python2 and wait for it to die.

Or someone check sphinx_doc_src/fish_indent_lexer.py, since I'm pretty sure that's where the problem happens.

**Comment by mqudsi** [comment]
Can you explicitly launch Sphinx via python3 (eg `python3 path/to/sphinx-build`) where that path would normally run via python2?

**Comment by faho** [comment]
Not really. I mean sphinx-build on my system looks like

```python
#!/usr/bin/python
# EASY-INSTALL-ENTRY-SCRIPT: 'Sphinx==2.0.1','console_scripts','sphinx-build'
__requires__ = 'Sphinx==2.0.1'
import re
import sys
from pkg_resources import load_entry_point

if __name__ == '__main__':
    sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])
    sys.exit(
        load_entry_point('Sphinx==2.0.1', 'console_scripts', 'sphinx-build')()
    )
```

but just running it via `python2 /usr/bin/sphinx-build` will error on the `pkg_resources` import, because that's installed in /usr/lib/python3.7/site-packages. So you'd have to mess with env variables to get that to work, and....

(To be clear, that's how switching from 3 to 2 looks, but it's equivalent the other way around)

-----

So python2 now _works_, for now. Personally, I'd be okay with keeping it this way until next year, when upstream py2 support ends, and then starting to require py3 for the optional features that we use python for.

**Comment by zanchey** [comment]
That's definitely going to bite us in the future; I'll take a look at the parser at some stage.

**Comment by zanchey** [comment]
Fixed with 2ca1bc43, though I haven't put the arrows back.

**Comment by faho** [comment]
> I haven't put the arrows back.

No need to. I found them a bit tacky to begin with (though for arrow _keys_ they're okay, as long as we can't guarantee a better display).
